src/run_graph.rs

```diff
@@ -79,66 +78,106 @@ impl TasksGraph {
         id
     }
 
-    pub fn next_task(&mut self) -> Option<(NodeIndex, Task)> {
+    pub fn next_task<DB: WriteResults>(
+        &mut self,
+        ex: &Experiment,
+        db: &DB,
+    ) -> Option<(NodeIndex, Arc<Task>)> {
         let root = self.root;
-        if let WalkResult::Task(id, task) = self.walk_graph(root) {
+        if let WalkResult::Task(id, task) = self.walk_graph(root, ex, db) {
             Some((id, task))
         } else {
```

Comment:
It seems to me that (today) the only two permissible variants here are `Task` and `NotBlocked` - `Blocked` indicates that graph dependencies have become nonlinear (possible at some point in the future) and something needs to change at the higher level to ensure we don't lose multithreading due to thread early exit.

What if we explicitly matched `Blocked` here to panic and say "If this panics, you've changed dependencies to become nonlinear and you need to change the top level threading behaviour to make sure threads don't exit early if they find themselves blocked". Just don't want to be tracking down strange slowdowns down the line.